torch/functional.py

```diff
@@ -102,12 +102,30 @@ def broadcast_shapes(*shapes):
     """
     # This wrapper exists to support variadic args.
     # TODO Movie this to C++ once the jit has better support for torch.Size.
-    with torch.no_grad():
-        scalar = torch.zeros((), device="cpu")
-        tensors = [scalar.expand(shape) for shape in shapes]
-        tensors = broadcast_tensors(*tensors)
-        return tensors[0].shape
-
+    max_len = 0
+    for shape in shapes:
+        if isinstance(shape, int):
+            if max_len < 1:
+                max_len = 1
+        elif isinstance(shape, tuple) or isinstance(shape, list):
+            s = len(shape)
+            if max_len < s:
+                max_len = s
+    result = [1] * max_len
+    for shape in shapes:
+        if isinstance(shape, int):
+            shape = (shape,)
+        if isinstance(shape, tuple) or isinstance(shape, list):
+            for i in range(-1, -1 - len(shape), -1):
+                if shape[i] < 0:
+                    raise RuntimeError("Trying to create tensor with negative dimension ({}): ({})"
+                                       .format(shape[i], shape[i]))
+                if shape[i] == 1 or shape[i] == result[i]:
+                    continue
+                if result[i] != 1:
+                    raise RuntimeError("Shape mismatch: objects cannot be broadcast to a single shape")
+                result[i] = shape[i]
```

Comment:
also, since you'll have to change this PR anyway, can you please add `else` here that will through an error for invalid input types (that are not integers or tuples or lists)

Reply:
Sure, I will update it soon. Thanks

Reply:
Updated :)

Reply:
Thanks, it makes sense to update.

#### Before Updating
```python
>>> import torch
>>> torch.broadcast_shapes(1.1)
torch.Size([])
>>> torch.broadcast_shapes((1.1, 1.2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/khushi/Documents/opensource/pytorch/torch/functional.py", line 132, in broadcast_shapes
    return torch.Size(result)
TypeError: torch.Size() takes an iterable of 'int' (item 0 is 'float')
>>> torch.broadcast_shapes((1.1, 1.2), (1.1, 1.2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/khushi/Documents/opensource/pytorch/torch/functional.py", line 132, in broadcast_shapes
    return torch.Size(result)
TypeError: torch.Size() takes an iterable of 'int' (item 0 is 'float')
>>> 
```

#### After Updating
```python
>>> import torch
>>> torch.broadcast_shapes((1.1, 1.2), (1.1, 1.2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/khushi/Documents/opensource/pytorch/torch/functional.py", line 124, in broadcast_shapes
    raise RuntimeError("Input shapes should be of type ints, a tuple of ints, or a list of ints.")
RuntimeError: Input shapes should be of type ints, a tuple of ints, or a list of ints.
>>> torch.broadcast_shapes(1.1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/khushi/Documents/opensource/pytorch/torch/functional.py", line 131, in broadcast_shapes
    raise RuntimeError("Input shapes should be of type ints, a tuple of ints, or a list of ints/")
RuntimeError: Input shapes should be of type ints, a tuple of ints, or a list of ints/
>>> 
```

#### NumPy's Output
```python
>>> import numpy
>>> numpy.broadcast_shapes(1.1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/khushi/anaconda3/envs/torch-dev/lib/python3.10/site-packages/numpy/lib/stride_tricks.py", line 471, in broadcast_shapes
    arrays = [np.empty(x, dtype=[]) for x in args]
  File "/home/khushi/anaconda3/envs/torch-dev/lib/python3.10/site-packages/numpy/lib/stride_tricks.py", line 471, in <listcomp>
    arrays = [np.empty(x, dtype=[]) for x in args]
TypeError: 'float' object cannot be interpreted as an integer
>>> numpy.broadcast_shapes((1.1, 1.2), (1.1, 1.2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/khushi/anaconda3/envs/torch-dev/lib/python3.10/site-packages/numpy/lib/stride_tricks.py", line 471, in broadcast_shapes
    arrays = [np.empty(x, dtype=[]) for x in args]
  File "/home/khushi/anaconda3/envs/torch-dev/lib/python3.10/site-packages/numpy/lib/stride_tricks.py", line 471, in <listcomp>
    arrays = [np.empty(x, dtype=[]) for x in args]
TypeError: 'float' object cannot be interpreted as an integer
```

I wanted to show this as these updates are failing in some cases at `test/test_distributions.py`. What do you think about it? Thanks